lib/concurrent/executor/executor_service.rb

```diff
@@ -0,0 +1,330 @@
+    #     must be called before this method (or on another thread).
+    #
+    #   @param [Integer] timeout the maximum number of seconds to wait for shutdown to complete
+    #
+    #   @return [Boolean] `true` if shutdown complete or false on `timeout`
+    def wait_for_termination(timeout = nil)
+      raise NotImplementedError
+    end
+
+    def running?
+      synchronize { ns_running? }
+    end
+
+    def shuttingdown?
+      synchronize { ns_shuttingdown? }
+    end
+
+    def shutdown?
+      synchronize { ns_shutdown? }
+    end
+
+    def auto_terminate?
+      synchronize { ns_auto_terminate? }
+    end
+
+    def auto_terminate=(value)
+      synchronize { self.ns_auto_terminate = value }
+    end
+
+    protected
+
+    # Handler which executes the `fallback_policy` once the queue size
+    # reaches `max_queue`.
+    #
+    # @param [Array] args the arguments to the task which is being handled.
+    #
+    # @!visibility private
+    def handle_fallback(*args)
+      case fallback_policy
+      when :abort
+        raise RejectedExecutionError
+      when :discard
+        false
+      when :caller_runs
+        begin
+          yield(*args)
+        rescue => ex
+          # let it fail
+          log DEBUG, ex
+        end
+        true
+      else
+        fail "Unknown fallback policy #{fallback_policy}"
+      end
+    end
+
+    def execute(*args, &task)
+      raise NotImplementedError
+    end
+
+    # @!macro [attach] executor_service_method_shutdown_execution
+    #
+    #   Callback method called when an orderly shutdown has completed.
+    #   The default behavior is to signal all waiting threads.
+    def shutdown_execution
+      # do nothing
+    end
+
+    # @!macro [attach] executor_service_method_kill_execution
+    #
+    #   Callback method called when the executor has been killed.
+    #   The default behavior is to do nothing.
+    def kill_execution
+      # do nothing
+    end
+
+    protected
+
+    def ns_auto_terminate?
+      !!@auto_terminate
+    end
+
+    def ns_auto_terminate=(value)
+      case value
+      when true
+        AtExit.add(self) { terminate_at_exit }
+        @auto_terminate = true
+      when false
+        AtExit.delete(self)
+        @auto_terminate = false
+      else
+        raise ArgumentError
+      end
+    end
+
+    def terminate_at_exit
+      kill # TODO be gentle first
+      wait_for_termination(10)
+    end
+  end
+
+  class RubyExecutorService < AbstractExecutorService
+
+    def initialize(*args, &block)
+      super
+      @stop_event    = Event.new
+      @stopped_event = Event.new
+      ensure_ivar_visibility!
+    end
+
+    def post(*args, &task)
+      raise ArgumentError.new('no block given') unless block_given?
+      synchronize do
+        # If the executor is shut down, reject this task
+        return handle_fallback(*args, &task) unless running?
+        execute(*args, &task)
```

Comment:
The `execute` method isn't very well named. This method is responsible for putting the task into the work queue. The `post` method checks for valid arguments and handles overflow when necessary. These operations are identical for all subclasses so we DRY them up but putting `post` in the base class. Each subclass has a different work queue so `execute` is delegated to the subclasses. Both `handle_fallback` and `execute` are synchronized because only one job can be placed into the work queue at any given time.